### PR TITLE
Feature/setup rpc api

### LINKED ISSUE #11: Add RPC call
Add RPC call interface for querying the following:
* Current balance of an account (free + reserved)
* Actual interest return per annum

Run a local testnet and test the RPC call.

### PR DESCRIPTION
**Pull request by IRISxiaowang**
Close #11 

Added to functions in rpc api
fn interest_pa()
fn fund_unlock_at()

Added low-level functions and tests in the bank pallet.

### CODE REVIEW

File: custom-rpc/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+publish = false
+
+[lints]
+workspace = true
+
+[dependencies]
+xy-chain-runtime = { path = '../runtime' }
+
+anyhow = "1.0"
+futures = "0.3.14"
```
Comment (syan095) [inline custom-rpc/Cargo.toml]: we should remove omse un-used dependencies

File: pallets/pallet-bank/src/lib.rs [hunk 2]
```diff
@@ -708,4 +757,37 @@ impl<T: Config> Pallet<T> {
 			}
 		})
 	}
+
+	// Return unlock block number.
+	pub fn fund_unlock_at(who: T::AccountId, lock_id: LockId) -> BlockNumberFor<T> {
+		AccountWithUnlockedFund::<T>::iter()
+			.find_map(|(block_number, accounts)| {
```
Comment (syan095) [inline pallets/pallet-bank/src/lib.rs]: I think find_map does a map (like a transformation) then a find.
Here we can just use find directly with the map. 
find(|(block, accounts)| accounts.contains)

File: pallets/pallet-bank/src/lib.rs [hunk 3]
```diff
@@ -708,4 +757,37 @@ impl<T: Config> Pallet<T> {
...
+					}
+				})
+			})
+			.unwrap_or_default()
+	}
+
+	pub fn interest_pa(who: T::AccountId) -> T::Balance {
+		let initial_balance = Self::accounts(who).reserved;
+		let interest_rate_per_year = Self::interest_rate();
+		let payout_times = FixedU128::saturating_from_rational(
```
Comment (syan095) [inline pallets/pallet-bank/src/lib.rs:777]: best to add more comments to make the calculation + type conversions more clear

Comment (syan095) [reply pallets/pallet-bank/src/lib.rs:777]: And add some spaces to break the method up into sections

File: pallets/pallet-bank/src/lib.rs (hunk 2 above)
Comment (syan095) [reply pallets/pallet-bank/src/lib.rs]: Also add documentations above the new functions

File: pallets/pallet-bank/src/tests.rs [hunk 4]
```diff
@@ -601,3 +610,46 @@ fn can_force_transfer() {
 			assert!(Bank::check_total_issuance());
 		});
 }
+
+#[test]
+fn can_check_fund_unlock_at() {
+	MockGenesisConfig::with_balances(vec![(ALICE, 1_000)]).build().execute_with(|| {
+		//
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs]: incomplete comment

File: pallets/pallet-bank/src/tests.rs [hunk 5]
```diff
@@ -601,3 +610,46 @@ fn can_force_transfer() {
...
+		let unlock_block = System::block_number() + StakePeriod::get();
+		assert_ok!(Bank::stake_funds(RuntimeOrigin::signed(ALICE), 100));
+
+		System::set_block_number(10);
+		let unlock_block_2 = System::block_number() + StakePeriod::get();
+		assert_ok!(Bank::stake_funds(RuntimeOrigin::signed(ALICE), 200));
+
+		System::set_block_number(100);
+		let unlock_block_3 = System::block_number() + StakePeriod::get();
+		assert_ok!(Bank::stake_funds(RuntimeOrigin::signed(ALICE), 300));
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs]: you can insert storage directly into the storage.
Test the case where there are multiple storage items, where each contains more than 1 locks - to ensure the search works.

File: pallets/pallet-bank/src/tests.rs [hunk 6]
```diff
@@ -601,3 +610,46 @@ fn can_force_transfer() {
...
+		assert_eq!(unlock_block, Bank::fund_unlock_at(ALICE, 1));
+		assert_eq!(unlock_block_2, Bank::fund_unlock_at(ALICE, 2));
+		assert_eq!(unlock_block_3, Bank::fund_unlock_at(ALICE, 3));
+	});
+}
+
+#[test]
+fn can_calculate_interest_pa() {
+	MockGenesisConfig::with_balances(vec![(ALICE, 1_000)]).build().execute_with(|| {
+		let initial_balance: Balance = 1_000_000_000_000_000_000;
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs]: initialize numbers as u128 to prevent an extra conversion and type declaration

File: pallets/pallet-bank/src/tests.rs [hunk 7]
```diff
@@ -601,3 +610,46 @@ fn can_force_transfer() {
...
+		let payout_times = YEAR / INTEREST_PAYOUT_PERIOD as u32;
+		Accounts::<Runtime>::mutate(ALICE, |account_data| account_data.reserved = initial_balance);
+
+		InterestRate::<Runtime>::set(Perbill::from_rational(10u32, 100u32));
+
+		let expect_interest = Bank::interest_pa(ALICE);
+
+		for _ in 0..payout_times {
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs]: add comments.
Here we are simulating the real interest as it is accumulated through the year

File: pallets/pallet-bank/src/tests.rs [hunk 8]
```diff
@@ -601,3 +610,46 @@ fn can_force_transfer() {
...
+			Bank::on_finalize(INTEREST_PAYOUT_PERIOD);
+		}
+
+		let actual_interest = Bank::accounts(ALICE).reserved - initial_balance;
+
+		assert_eq_with_precision(expect_interest, actual_interest, 1_000_000u128);
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs]: are we only accurate to 6 digits?

Comment (syan095) [reply pallets/pallet-bank/src/tests.rs]: Test the case where actual interest rate can go above 1 (like 100% interest rate).

You can write the interest simulate + calculation as a test function, then call the test multiple times with different interest rate.

Test cases where the interest rate is 0%, 100% etc. to cover edge cases

**Review by syan095 — CHANGES_REQUESTED** [review]
Need to brush up on the details. Ensure the code is to a high standard.
(this includes things like formatting, variable names, comments and documentations)

Also, the tests need to cover more cases. The general rule is that the unit tests need to test at least the edge cases, 1 good pathway and 1 bad pathway

File: .vscode/settings.json [hunk 9]
```diff
@@ -1,3 +1,6 @@
 {
```
Comment (syan095) [inline .vscode/settings.json]: THis file should be gitignored

File: pallets/pallet-bank/src/tests.rs [hunk 10]
```diff
@@ -601,3 +603,77 @@ fn can_force_transfer() {
...
+	MockGenesisConfig::with_balances(vec![(ALICE, 1_000)]).build().execute_with(|| {
+		// Check if it works with stake function.
+		let unlock_block = System::block_number() + StakePeriod::get();
+		assert_ok!(Bank::stake_funds(RuntimeOrigin::signed(ALICE), 100));
+
+		// Verify
+		assert_eq!(unlock_block, Bank::fund_unlock_at(ALICE, 1));
+
+		// Check it works with `AccountWithUnlockedFund` stores with multiple lock id.
+		AccountWithUnlockedFund::<Runtime>::insert(10, vec![(ALICE, 2), (ALICE, 5), (ALICE, 6)]);
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs]: you need to add more test data - more users and a mixture of users and lock ID s to fully test the search function.

1 block with fully mixed locks from 3 users
blocks with locks from another user with the same lock ID

Comment (syan095) [reply pallets/pallet-bank/src/tests.rs]: Also test the case the lock cannot be found.

File: pallets/pallet-bank/src/tests.rs [hunk 11]
```diff
@@ -601,3 +603,77 @@ fn can_force_transfer() {
...
+
+		calculate_interest_pa_with_interest_rate(initial_balance, 10u32);
+
+		calculate_interest_pa_with_interest_rate(initial_balance, 9_999u32);
+
+		calculate_interest_pa_with_interest_rate(initial_balance, 10_000u32);
+	});
+}
+
+// Test interest can be received per year by using Alice account with different interest rate and
```
Comment (syan095) [inline pallets/pallet-bank/src/tests.rs]: documentations above functions should use ///
comments within functions use //

**Review by syan095 — APPROVED** [review]
LGTM
